compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1345,7 +1345,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     }
 
     def isScalaPhantomSelf(self: tpd.ValDef): Boolean = {
-      if (self.tpt.tpe.derivesFrom(defn.PhantomClass)) {
+      if (self.tpt.tpe.classSymbol eq defn.PhantomClass) {
```

Comment:
Can you explain what the restriction is supposed to prevent? I did not get why it's `eq PhantonClass`, or even why having PhantomClass as a selftype is a problem. I would think it more plausible that phantom _types_, i.e. members of a subclass of `PhantomClass` should not be self types, but even there I would need to see detailed justification. It's true that classes with such self types cannot be instantiated, but why enforce that at the selftype level?